### ISSUE TITLE
Ambiguous use of tune in samplers

### ISSUE THREAD

**Issue by axsk** [brian-j-smith/Mamba.jl#66]
Currently many samplers (amm, hmc, mala, ..) use two different forms and locations to store their tune parameters:
- Sampler.tune of type Dict
- Sampler.tune["sampler"] of some sampler-specific tune type

Some parameters are stored in both, implying forth and back copying.
This seems redundant and confusing to me. Is there any reason for that?

I think the cleanest way would be to introduce an `abstract Tune`, let the `tune` field of `Sampler` be of that type, and just use the sampler specific tune types `SamplerTune <: Tune`.

On the way there we could use Sampler.tune :: Union{Dict,Tune} for compability to yet unported samplers.

What do you think of this?

**Comment by brian-j-smith** [comment]
I'll try to explain some of the issues involved.  For those reading this post, do note that an understanding of the following details is not needed to use samplers out of the box.  This will mainly be of interest to developers implementing new samplers or wanting to understand the implementation of existing ones.

Some of the issues...

1.  Where it makes sense to do so, stand-alone samplers are provided for use apart from the ``Model`` specification syntax.  ``amm!()``, ``hmc!()``, and ``mala!()`` are examples of such samplers.
2.  Stand-alone-sampler-specific types are defined to store tuning parameters that need to be saved and used internally from iteration to iteration.
3.  Stand-alone samplers can be linked to the ``Model`` syntax with the ``Sampler`` type.  The latter serves as a wrapper for the linkage.
4.  The ``Sampler.tune`` field is a dictionary type for storing data that are needed in calls to a stand-alone sampler.  In general, the data consist of user-supplied arguments to the Sampler-based constructor and stand-alone-sampler-specific tuning parameters.  (The dictionary entries can be defined and used as a developer sees fit.)
5.  IMO, it is probably better to not combine the user-supplied and sampler-specific data into one ``Tune`` subtype because 1) the ``Sampler`` type and stand-alone function serve different roles, and 2) the stand-alone function is intended to be used apart from the ``Sampler`` type and its user-supplied data.

If someone was doing sampler development and made a really compelling case for a different type in the ``Sampler.tune`` field, then I would be open to something like defining an abstract ``AbstractSampler`` type of which the existing ``Sampler`` would be a concrete type and other subtypes could be defined.

Regardless, there will likely be some tweaks to the Sampler framework, so suggestions are welcome.

P.S.  As it currently stand, there is not much back and forth copying of tuning parameters.  Since julia uses a "pass-by-sharing" model, it is primarily just new variable bindings being created rather than underlying data being copied.

**Comment by brian-j-smith** [comment]
@axsk - Over the course of making updates to the Sampler-based constructors this week, the tuning parameter structures were generalized in a way that addresses your original post.  In particular, ``SamplerTune`` was defined as an abstract type from which concrete tune types descend, and ``Sampler`` was redefined with a variable type for the ``tune`` field as shown below.  Consequently, a developer will be able to use any type to store tuning parameters.  These changes are in the master branch and are planned to be included in the next package release (0.7.4).

```julia
  ## Type definitions
  abstract SamplerTune

  type Sampler{T}
    params::Vector{Symbol}
    eval::Function
    tune::T
    targets::Vector{Symbol}
  end

  # Constructor
  Sampler(params::Vector{Symbol}, f::Function, tune::Any=Dict())
```

**Comment by axsk** [comment]
In order to resume a mcmc simulation which I store on disk in-between I tried overwriting the model.samplers[1].tune field after I ```setsamplers!```. Unfortunately a call to SamplerVariate overwrites this [on the first iteration](https://github.com/brian-j-smith/Mamba.jl/blob/81cecc02a7d525c4dc2482845b6c6447223ef06a/src/samplers/sampler.jl#L25).
Is this actually necessary, as usually the sampler initialization already specifies a default tune field.
This looks like duplicate code to me, maybe we can just check whether s.tune is already defined instead of checking for the first iteration?

**Comment by axsk** [comment]
I am also not sure on the design saving the samplers/their tune in the model.
When computing multiple chains, only the model of the first chain is saved back to the ModelChains object. This also implies only the first tune field is saved back to the model. Continuation of the chains will all use that tune field, which probably is not totally correct.

I think a clearer separation might clean up the code and fix this problem, and beeing a fan of immutable structures (the above post is an example why mutability complexes things), here is my (work in progress) proposal of separation:
 - Model: just the (immutable) model, without any state/tune/values,
 - SamplingScheme beeing an array of (immutable) Sampler objects
 - ModelState, used for computation/update/...
 - Chain:
  - used for the init values
  - containing fields to for model, samplingscheme and current tune fields

This would specificalliy simplify the mcmc methods, as the treatment of start and continuation would collapse, treatment of multiple chains would simplify of just (p)mapping "initial chains" to mcmc.

I haven't thought this through completely yet, so consider this just an idea to think about.

**Comment by brian-j-smith** [comment]
There is a straightforward change in the master branch now for storing and restoring tuning parameters for restarting multiple chains.

With respect to your previous post about storing mcmc simulations in external files for future julia session, I would suggest using the ``serialize/deserialize`` julia functions for that (see example code below).  There is another commit in master that is needed for this to fully work in multiprocessor mode and that will be in the next package release (0.8.1).

Otherwise, *Mamba* is intended to be used through its exported constructors and method functions.  Object fields, like ``model.samplers[1].tune``, should not be changed manually by users.

```julia
## assume mcmc results are saved in sim
## write sim to an external file
serialize(open("sim.jls", "w"), sim)

## read sim after restarting julia
using Mamba
sim = deserialize(open("sim.jls", "r"))

## restart the sampler
sim2 = mcmc(sim, 1000)
```

**Comment by axsk** [comment]
Although late, here I am back on the (kind of evolved) topic.
First, thank you for the changes so far.

I agree that direct usage of internal fields should be discouraged, so the ```settune!``` function is reasonable.

But I think it still leads to unexpected behaviour in a code like
``` julia
m = Model(...)
setsamplers!(m, SomeSampler(...))
settune!(m, customtune, 0)
sim = mcmc(m, ...)
```
as the tune field gets overwritten [on the first iteration](https://github.com/brian-j-smith/Mamba.jl/blob/master/src/samplers/sampler.jl#L40).

As in my use case the chains become very large I actually use such a construction to continue/read in only the last iteration, using my own serialization via JLD.

A way to just check if the tune field was already initialized independently of the iteration would be to replace
``` julia
if iter == 1
``` 
by 
``` julia
if any(map(f -> !isdefined(s.tune, f), fieldnames(s.tune)))
```

**Comment by brian-j-smith** [comment]
* Good to hear that you like the changes.  Some of the internal tuning parameters that you had mentioned wanting to change (e.g. AMMTune.scale) are now accessible through the sampler constructors.
* ``settune!()`` is not an exported (or documented) function and, as such, is for internal use only.  Anyone manually importing and using it should not have expectations for it beyond their own understanding of the internal workings of the package.
* Tuning parameters are reset at the first iteration (``iter == 1``) of each chain, as they should be to ensure independence of chains.  Your proposed coding change does not preserve that behavior and could allow tuning parameters changed (e.g. adaptively) in one chain to be passed on to other chains.  The original behavior needs to be preserved.
* Restarting of chains has been available in the package through one of the ``mcmc()`` methods for a while now.  I consider that to be proof that such a task is possible with the current code base.  In other words, I do not see why coding changes would be needed to restart chains in your case.
* A fundamental issue in what you have described above is that the ``iter`` field is being set to 0 when restarting samplers.  In your custom algorithm, you should advance the value of that field to the iteration from which a chain is being restarted.  That is what the ``mcmc()`` method does.  Once you get the iteration number set correctly, and address any related issues, my guess is things will work fine.
* The changes in pull request #85 should not be needed to get this working.  Plus, some of them are code breaking.

**Comment by axsk** [comment]
@3.: Actually the reset at the first iteration also doesn't ensure independence of the chains. It basically has the same problem as my proposal, consider following progression:
`chain 1 iter 1 -> chain 2 iter 1 -> chain 1 iter 2 -> chain 2 iter 2`
The current implementation resets the tuning parameter in the first 2 steps, but then the parameter is carried over from chain 2 to chain 1 and so on, as there is only one instance of a sampler, thus only one tune field.
My code suggestion just resets (actually initializes, if not yet initialized) once, which I think is at least consistent with behaviour on following iterations (no resets).

@4.: Let me clarify my goal: I dont want to restart the chain, but continue from just the last given sample (as initial condition, since I cannot load the whole chain into memory) using the (customly) stored adaption from the tunefield.
This is akin to starting a chain with custom tune/adaption information, a in my eyes sensible operation, for which
``` julia 
m = Model(...)
setsamplers!(m, SomeSampler(...))
settune!(m, customtune, 0)
sim = mcmc(m, Dict{Symbol}(), [Dict(:x=>load_start_value())], iters)
``` 
looks like the right semantics.

@5: In this setting the iteration number indeed should be 0 in the beginning.

@6: The code breaking is bad of course, I will fix that asap. 
But I still am convinced that the general idea is the correct/clean way.

**Comment by axsk** [comment]
I think CI should pass now.
BUT, I misunderstood how mcmc worked.
So I want to revise what I wrote abound the independence for now.

What I am actually wondering is how the whole parallel process actually works.
If in the pmap we have the _same_ model for each chain, then inject [(relist!, settune!)](https://github.com/axsk/Mamba.jl/blob/settune/src/model/mcmc.jl#L65) the values into it and sample concurrently, how can this be uncorrelated?

**Comment by axsk** [comment]
Ok I hope I got it now.
Here we pass the very same model m (if `nworkers()==1`) to the mutating function mcmc_worker!, but reset the (mutating) tune field with the setinits!(::Model, ::Vector{Dict}) for each chain seperately to a deepcopy, i.e. we end up with different tune fields for each chain. 
Thus my proposa PR also resets the tune field for each chain.
For evidence see [here](https://gist.github.com/axsk/4c32dcd0c7a90a483077)

**Comment by brian-j-smith** [comment]
By "restarting of chains" I mean restarting chains at the last iterations.  As I said, a version of this has been available in Mamba for a while (since 2014).  So, it is doable with the current code base.  I initially thought your "continuation" meant the same thing, but the implementation you describe is something different.  So, I am confused now about what you want to do.

If I understand correctly, "continuation" to you means extending a chain from its last iteration and using a value of 1 for the next iteration number.  As you probably know, an MCMC chain is a time series.  Iteration numbers locate MCMC draws in the time series from which they come.  Thus, iteration numbers have meaning - iteration 0 means the start of a new MCMC chain, and smaller iteration numbers represent earlier draws in a chain.  So, are you extending a chain, in which case the iteration number should be incremented from where it left off; or are you starting a new chain, starting with iteration 0?  If the latter, then note that you are effectively carrying tuning parameters from an existing chain over to a new one (not independent).

The above may sound like a quibble over semantics, but iteration numbers have real implications for the package.  For instance, tuning parameters are set at the start of a new chain (``iter==1``), burn-in and any associated adaptive tuning is applied at the start (``1:burnin``), and chains are concatenated according to their iteration numbers.  All of these things are issues if a chain is started at iteration 0 when it was really meant to be a continuation of an existing chain.

With respect to the PR implementation, I have some reservations, including the following.
1. Tuning parameter values passed in to Sampler constructors will be ignored in certain circumstances.  That a constructor would not use values passed to it to instantiate its object seems like a distortion of the purpose of a constructor.
2. Increased chances for user error when tuning parameters are specified apart from the constructor.  For example, no checks would be performed to ensure that tuning parameter and variate spaces are of the same dimensions.
3. Initialization additionally requires tuning parameters to have undefined fields.  This may not affect current package samplers, but could affect other people writing samplers and unaware of the additional requirement.
4. More computation needed, at every iteration, for the initialization checks.

A more transparent and explicit way of constructing a Sampler from a tune object would be to define a constructor specifically for that purpose.  Such an approach would also address my reservations above.  Below is an example of what you could implement without needing to change the package.

```julia
function AMM(params::ElementOrVector{Symbol}, tune::AMMTune; adapt::Symbol=:all)
  adapt in [:all, :burnin, :none] ||
    throw(ArgumentError("adapt must be one of :all, :burnin, or :none"))

  samplerfx = function(model::Model, block::Integer)
    block = SamplingBlock(model, block, true)
    v = SamplerVariate(block, tune)
    isadapt = adapt == :burnin ? model.iter <= model.burnin :
              adapt == :all ? true : false
    sample!(v, x -> logpdf!(block, x), adapt=isadapt)
    relist(block, v)
  end
  Sampler(params, samplerfx, AMMTune())
end
```

In the package, the implementation might look something like...

```julia
function AMM{T<:Real}(params::ElementOrVector{Symbol}, Sigma::Matrix{T};
                      args...)
  AMMSampler(params, Sigma; args...)
end

function AMM(params::ElementOrVector{Symbol}, tune::AMMTune; args...)
  AMMSampler(params, tune; args...)
end

function AMMSampler(params, pargs...; adapt::Symbol=:all, kargs...)
  adapt in [:all, :burnin, :none] ||
    throw(ArgumentError("adapt must be one of :all, :burnin, or :none"))

  samplerfx = function(model::Model, block::Integer)
    block = SamplingBlock(model, block, true)
    v = SamplerVariate(block, pargs...; kargs...)
    isadapt = adapt == :burnin ? model.iter <= model.burnin :
              adapt == :all ? true : false
    sample!(v, x -> logpdf!(block, x), adapt=isadapt)
    relist(block, v)
  end
  Sampler(params, samplerfx, AMMTune())
end
```

Now, I am not endorsing the idea of using tuning parameters from one chain to start another; but if one were to do that, this is the approach I would recommend.  Otherwise, I think you want to be starting chains and iterations from where they left off.

**Comment by axsk** [comment]
Thank you for your detailed answer.
I now agree with all your objections :+1: 

I guess defining that sampler as above would fix my problem, though I do not know anymore if it is the right approach.

The main reason why I don't use the write/read functions to continue my chain is (besides version compability, which I did not test that way but I read the serialize/deserialize are sensitive to version/sysimg, and I am using a Cluster with different architectures) that I do not want to hold the whole chain in memory.
I compute my chain in a batch workflow, computing 100.000 samples, saving them as an Array{Float64} in an JLD (which supports growing files), then reading in the first, computing the next 100.000, save them and so on (with restart of Julia instances inbetween).

I hoped to be able to just use the settune! function to carry over the tune field to a new chain using the last sample as inits.
But that field got overwritten on the first iteration.
Also manually setting the model.iter field gets overwritten by setinits! (and I should not access Mamba internal fields directly !! :)).
So I tried creating a ```ModelChains(Chain(lastiter), newmodel())``` but this fails as the (after Julia restart) new model has no ModelState.
I could try storing the ModelChains minus the values, but serialize might induce problems with different Julia versions and the last time I tried using JLD for this [failed](https://github.com/brian-j-smith/Mamba.jl/issues/67) which is why I resorted to creating the model anew each time.

TL;DR: How would you continue the chain when storing the ```chains.value``` separately as Array{Float64} and reading in just the last value, without resetting the Samplers tune information.

**Comment by brian-j-smith** [comment]
In the setting described above where individual iterations are saved separately with JLD and the goal is to continue chains, I would use the serialize-based write/read methods to save only the last iterations of corresponding ``ModelChains`` objects (see below) - until or unless something causes the serialize approach to fail.

```julia
## Write simulation state
sim = mcmc(...)
write("sim_last.jls", sim[end:end, :, :])

## Read state and continue chain
sim_last = read("sim_last.jls", ModelChains)
sim2 = mcmc(sim_last, 100000)
```

**Comment by axsk** [comment]
That absolutely makes sense and now works like a charm, thank you!
I guess closing this issue (+ q&a) is overdue, I hope that's ok.

**Comment by brian-j-smith** [comment]
Thanks Alex.

**Comment by axsk** [comment]
Guess I was happy too fast.
Using `serialize`/`deserialize` randomly throws segfaults in my code :(
At least I am not the only one with the problem :)
https://github.com/JuliaLang/julia/issues/14497#issuecomment-195602944

I will probably just live without the inbetween saves/analysis and wait some days before examining my results :(

**Comment by brian-j-smith** [comment]
Here's hoping the julia developers can find/fix the serialize issue soon.
